File: lua/binary/binary.lua
local function to_decimal(input)
	local result = 0

	if not input:match("^([01]+)$") then
		return result
	end

	local reversed = input:reverse()
	for i = 1, #reversed do
		local n = tonumber(reversed:sub(i, i))

		result = result + 2 ^ (i - 1) * n
	end

	return result
end

return {
	to_decimal = to_decimal,
}

File: c/binary/binary.c
#include "binary.h"

int convert(const char *input) {
  int result = 0;

  for (const char *it = input; *it != '\0'; it++) {
    int digit = (*it - '0');

    if (digit != 0 && digit != 1) {
      return INVALID;
    }

    result = result << 1 | digit;
  }

  return result;
}
